### ISSUE TITLE
bag.groupby doesn't work properly with string keys

### ISSUE THREAD

**Issue by blokhinnv** [dask/dask#6840]
I'm going through the [the second part of the tutorial](https://github.com/dask/dask-tutorial/blob/master/02_bag.ipynb) and it looks like `dask.bag.groupby` doesn't work properly.

```
####### the first part is a slightly modified version of the code from prep.py
import numpy as np
import time
import pandas as pd
import os
import gzip

names = ['Alice', 'Bob', 'Charlie', 'Dan', 'Edith', 'Frank', 'George',
'Hannah', 'Ingrid', 'Jerry', 'Kevin', 'Laura', 'Michael', 'Norbert', 'Oliver',
'Patricia', 'Quinn', 'Ray', 'Sarah', 'Tim', 'Ursula', 'Victor', 'Wendy',
'Xavier', 'Yvonne', 'Zelda']

def account_params(k):
    ids = np.arange(k, dtype=int)
    names2 = np.random.choice(names, size=k, replace=True)
    wealth_mag = np.random.exponential(100, size=k)
    wealth_trend = np.random.normal(10, 10, size=k)
    freq = np.random.exponential(size=k)
    freq /= freq.sum()

    return ids, names2, wealth_mag, wealth_trend, freq

def json_entries(n, *args):
    df = account_entries(n, *args)
    g = df.groupby(df.id).groups

    data = []
    for k in g:
        sub = df.iloc[g[k]]
        d = dict(id=int(k), name=sub['names'].iloc[0],
                transactions=[{'transaction-id': int(i), 'amount': int(a)}
                              for i, a in list(zip(sub.index, sub.amount))])
        data.append(d)

    return data


def accounts_json(data_dir, small=False):
    t0 = time.time()

    if small:
        num_files, n, k = 50, 10000, 250
    else:
        num_files, n, k = 50, 100000, 500
        
    fn = os.path.join(data_dir, 'accounts.%02d.json.gz' % (num_files - 1))
    if os.path.exists(fn):
        return

    args = account_params(k)

    for i in range(num_files):
        seq = json_entries(n, *args)
        fn = os.path.join(data_dir, 'accounts.%02d.json.gz' % i)
        with gzip.open(fn, 'wb') as f:
            f.write(os.linesep.join(map(json.dumps, seq)).encode())

    t1 = time.time()
    print("Created JSON acccouts in {:0.2f}s".format(t1 - t0))
accounts_json('dask_data')
####### data preparation ended

import os
import dask.bag as db
import json
lines = db.read_text(os.path.join('dask_data', 'accounts.*.json.gz'))
js = lines.map(json.loads)

result = js.groupby(lambda item: item['name']).starmap(lambda k, v: (k, len(v))).compute()
print(sorted(result))
```
Here I expect that `result` list would contain only unique keys, but the output shows that there is four tuples for each key. 
```
[('Alice', 268), ('Alice', 270), ('Alice', 290), ('Alice', 292), ('Bob', 228), ('Bob', 228), ('Bob', 247), ('Bob', 247), ('Charlie', 192), ('Charlie', 192), ('Charlie', 208), ('Charlie', 208), ('Dan', 200), ('Dan', 201), ('Dan', 219), ('Dan', 219), ('Edith', 252), ('Edith', 252), ('Edith', 273), ('Edith', 273), ('Frank', 228), ('Frank', 228), ('Frank', 247), ('Frank', 247), ('George', 300), ('George', 325), ('George', 624), ('Hannah', 216), ('Hannah', 216), ('Hannah', 234), ('Hannah', 234), ('Ingrid', 140), ('Ingrid', 140), ('Ingrid', 150), ('Ingrid', 152), ('Jerry', 204), ('Jerry', 204), ('Jerry', 221), ('Jerry', 221), ('Kevin', 250), ('Kevin', 252), ('Kevin', 271), ('Kevin', 272), ('Laura', 252), ('Laura', 252), ('Laura', 273), ('Laura', 273), ('Michael', 180), ('Michael', 180), ('Michael', 195), ('Michael', 195), ('Norbert', 228), ('Norbert', 247), ('Norbert', 475), ('Oliver', 180), ('Oliver', 180), ('Oliver', 195), ('Oliver', 195), ('Patricia', 192), ('Patricia', 192), ('Patricia', 208), ('Patricia', 208), ('Quinn', 359), ('Quinn', 360), ('Quinn', 390), ('Quinn', 390), ('Ray', 262), ('Ray', 263), ('Ray', 284), ('Ray', 284), ('Sarah', 312), ('Sarah', 312), ('Sarah', 676), ('Tim', 240), ('Tim', 240), ('Tim', 260), ('Tim', 260), ('Ursula', 192), ('Ursula', 192), ('Ursula', 208), ('Ursula', 208), ('Victor', 228), ('Victor', 228), ('Victor', 247), ('Victor', 247), ('Wendy', 156), ('Wendy', 156), ('Wendy', 169), ('Wendy', 169), ('Xavier', 264), ('Xavier', 264), ('Xavier', 285), ('Xavier', 286), ('Yvonne', 245), ('Yvonne', 248), ('Yvonne', 266), ('Yvonne', 267), ('Zelda', 204), ('Zelda', 204), ('Zelda', 221), ('Zelda', 221)]
```
Why would it work that way? 

**Environment**:

- Dask version: 2.20.0
- Python version: 3.8.3
- Operating System: windows 10
- Install method (conda, pip, source): Anaconda

**Comment by blokhinnv** [comment]
I managed to fix this behaviour by adding 2 parameters in the call:
```
result = js.groupby(lambda item: item['name'], shuffle='disk', npartitions=1).starmap(lambda k, v: (k, len(v))).compute()
print(sorted(result))
```
So, how it looks from where I'm standing. `js` is split at about 50 partitions by `dask` when created. So is `result` before computations. After I call compute it looks like data are groupped only inside a partition. 
I can't find anything about such behaviour of `dask.bag.groupby` in docs. The tutorial says "groupby: Shuffles data so that all items with the same key are in the same key-value pair". Obviously it doesn't in my case. The behaviour I described above looks strange for me -- what is the point of groupping only inside partitions?
